contracts/interchain-token-service/tests/execute.rs

```diff
@@ -573,3 +552,411 @@ fn set_chain_config_should_fail_if_chain_config_is_already_set() {
+        message: InterchainTransfer {
+            token_id: token_id.clone(),
+            source_address: HexBinary::from([1; 32]).try_into().unwrap(),
+            destination_address: HexBinary::from([2; 32]).try_into().unwrap(),
+            amount: 1u64.try_into().unwrap(),
+            data: None,
+        }
+        .into(),
+    };
+    assert_err_contains!(
+        utils::execute_hub_message(
+            deps.as_mut(),
+            CrossChainId::new(
+                destination_its_chain.clone(),
+                router_message.cc_id.message_id.clone()
+            )
+            .unwrap(),
+            destination_its_contract.clone(),
+            msg,
+        ),
+        ExecuteError,
+        ExecuteError::TokenSupplyInvariantViolated { .. }
+    );
+
+    let msg = HubMessage::SendToHub {
+        destination_chain: destination_its_chain.clone(),
+        message: InterchainTransfer {
+            token_id: token_id.clone(),
+            source_address: HexBinary::from([1; 32]).try_into().unwrap(),
+            destination_address: HexBinary::from([2; 32]).try_into().unwrap(),
+            amount,
+            data: None,
+        }
+        .into(),
+    };
+    assert_ok!(utils::execute_hub_message(
+        deps.as_mut(),
+        router_message.cc_id.clone(),
+        source_its_contract.clone(),
+        msg,
+    ));
+
+    let msg = HubMessage::SendToHub {
+        destination_chain: source_its_chain.clone(),
+        message: InterchainTransfer {
+            token_id: token_id.clone(),
+            source_address: HexBinary::from([1; 32]).try_into().unwrap(),
+            destination_address: HexBinary::from([2; 32]).try_into().unwrap(),
+            amount: amount.strict_add(Uint256::one()).try_into().unwrap(),
+            data: None,
+        }
+        .into(),
+    };
+    assert_err_contains!(
+        utils::execute_hub_message(
+            deps.as_mut(),
+            CrossChainId::new(destination_its_chain, router_message.cc_id.message_id).unwrap(),
+            destination_its_contract,
+            msg,
+        ),
+        ExecuteError,
+        ExecuteError::TokenSupplyInvariantViolated { .. }
+    );
+}
+
+#[test]
+fn deploy_interchain_token_submitted_twice_fails() {
+    let (
+        mut deps,
+        TestMessage {
+            router_message,
+            source_its_contract,
+            hub_message: msg,
+            ..
+        },
+    ) = utils::setup();
+
+    assert_ok!(utils::execute_hub_message(
+        deps.as_mut(),
+        router_message.cc_id.clone(),
+        source_its_contract.clone(),
+        msg.clone(),
+    ));
+
+    assert_err_contains!(
+        utils::execute_hub_message(
+            deps.as_mut(),
+            router_message.cc_id.clone(),
+            source_its_contract.clone(),
+            msg,
+        ),
+        ExecuteError,
+        ExecuteError::TokenAlreadyDeployed { .. }
+    );
+}
+
+#[test]
+fn deploy_interchain_token_from_non_origin_chain_fails() {
```

Comment:
I didn't fix this test on purpose. The deployment fails as expected, but the reason is that there already is a token deployed on that chain. I believe a test that checks that you can't deploy from a non-origin chain should actually try to deploy to a chain without token.

Reply:
So this is up for discussion

Reply:
Yeah, the test was relying on the ordering of the checks so it would hit the origin chain check first and fail. But I agree it's better to use a new destination chain as well where the token is not deployed to make it more clear. It should fail in both cases, but the latter failure is more obviously a result of the check in question